distributed/shuffle/_rechunk.py

```diff
@@ -166,45 +169,310 @@ def rechunk_unpack(
+    old: NDSlice
+    #: n-dimensional slice of the new chunks along each axis that belong to the partial
+    new: NDSlice
+    #: Indices of the first value of the left-most old chunk along each axis
+    #: to include in this partial. Everything left to this index belongs to
+    #: the previous partial.
+    left_starts: NDIndex
+    #: Indices of the first value of the right-most old chunk along each axis
+    #: to exclude from this partial.
+    #: This corresponds to `left_start` of the subsequent partial.
+    right_stops: NDIndex
 
+
+def rechunk_p2p(x: da.Array, chunks: ChunkedAxes) -> da.Array:
     import dask.array as da
 
     if x.size == 0:
         # Special case for empty array, as the algorithm below does not behave correctly
         return da.empty(x.shape, chunks=chunks, dtype=x.dtype)
 
-    dsk: dict = {}
+    dsk = {}
     token = tokenize(x, chunks)
-    _barrier_key = barrier_key(ShuffleId(token))
-    name = f"rechunk-transfer-{token}"
+    name = f"rechunk-p2p-{token}"
+    for ndpartial in _split_partials(x, chunks):
+        n_new_chunks = math.prod(slc.stop - slc.start for slc in ndpartial.new)
+        if n_new_chunks == 1:
+            dsk.update(partial_concatenate(x, chunks, ndpartial, name))
+        else:
+            dsk.update(
+                partial_rechunk(
+                    x,
+                    chunks=chunks,
+                    ndpartial=ndpartial,
+                    name=name,
+                )
+            )
+    layer = MaterializedLayer(dsk)
+    graph = HighLevelGraph.from_collections(name, layer, dependencies=[x])
+    arr = da.Array(graph, name, chunks, meta=x)
+    return arr
+
+
+def _split_partials(
+    x: da.Array, chunks: ChunkedAxes
+) -> Generator[_NDPartial, None, None]:
+    """Split the rechunking into partials that can be performed independently."""
+    partials_per_axis = _split_partials_per_axis(x, chunks)
+    for partial_per_axis in product(*partials_per_axis):
+        old, new, left_starts, right_stops = zip(*partial_per_axis)
+        yield _NDPartial(old, new, left_starts, right_stops)
+
+
+def _split_partials_per_axis(
+    x: da.Array, chunks: ChunkedAxes
+) -> tuple[tuple[_Partial, ...], ...]:
+    """Split the rechunking into partials that can be performed independently
+    on each axis."""
+    from dask.array.rechunk import old_to_new
+
+    chunked_shape = tuple(len(axis) for axis in chunks)
+    _old_to_new = old_to_new(x.chunks, chunks)
+
+    sliced_axes = _partial_slices(_old_to_new, chunked_shape)
+
+    partial_axes = []
+    for axis_index, slices in enumerate(sliced_axes):
+        partials = []
+        for slice_ in slices:
+            last_old_chunk: int
+            first_old_chunk, first_old_slice = _old_to_new[axis_index][slice_.start][0]
+            last_old_chunk, last_old_slice = _old_to_new[axis_index][slice_.stop - 1][
+                -1
+            ]
```

Comment:
Slightly more pythonic and coherent with `_partial_slices`:
```suggestion
    for slices, old_to_new_axis in zip(sliced_axes, _old_to_new):
        partials = []
        for slice_ in slices:
            last_old_chunk: int
            first_old_chunk, first_old_slice = old_to_new_axis[slice_.start][0]
            last_old_chunk, last_old_slice = old_to_new_axis[slice_.stop - 1][-1]
```